KibanaTypeLogzio: Tests exceed Auth0 MFA rate limit

### Terraform Version

`v0.11.14`

### Affected Resource(s)

**Logzio** test cases. Normal operation appears unaffected.

### Terraform Configuration Files

N/A. Test cases in master branch of repository.

### Debug Output

```
--- FAIL: TestAccDataSourceKibanaIndex_Basic (8.50s)
    testing.go:564: Error destroying resource! WARNING: Dangling resources
        may exist. The full state and error is shown below.

        Error: Error refreshing: 1 error occurred:

        * data.kibana_index.basic: 1 error occurred:

        * data.kibana_index.basic: data.kibana_index.basic: could not get saved objects, error: [Error getting MFA code: the mfa code sent is expired]

        State: <nil>
```

### Expected Behavior

Test cases authenticate against Logz.io exactly once, exchanging the credentials and MFA token for a session token which is used throughout the tests.

### Actual Behavior

The output above is erroneous due to the logic in [ewilde/go-kibana logzio_authentication_handler.go#L206](https://github.com/ewilde/go-kibana/blob/master/logzio_authentication_handler.go#L206).

The real error is masked: 

```json
{
  "error": "a0.mfa_too_many_codes",
  "error_description": "Too many attempts. Please wait a few minutes before retrying."
}
```

`ConfigureProvider()` is called multiple times per test case. Specifically, every time `ctx.Refresh()`, `ctx.Plan()`, `ctx.Apply()`, and `ctx.Destroy()` are called. This results in us hammering the Auth0 API to the point of exceeding the rate limit before a single test case has completed.

### Steps to Reproduce

1. Configure the environment variables for an MFA-enabled `KibanaTypeLogzio` test environment.
2. `TF_ACC=1 go test -v ./kibana -run TestAccDataSourceKibanaIndex_Basic`

--- a/.travis.yml
+++ b/.travis.yml
@@ -12,7 +12,7 @@ after_success:
 env:
   matrix:
   - ELK_VERSION=5.5.3   KIBANA_TYPE=KibanaTypeVanilla   KIBANA_USERNAME=elastic             KIBANA_PASSWORD=changeme  ELASTIC_PACK=
-  - ELK_VERSION=5.5.3   KIBANA_TYPE=KibanaTypeLogzio    KIBANA_URI=https://app-eu.logz.io   ELASTIC_SEARCH_PATH=/kibana/elasticsearch/logzioCustomerKibanaIndex
+  - ELK_VERSION=6.3.2   KIBANA_TYPE=KibanaTypeLogzio    KIBANA_URI=https://app-eu.logz.io   ELASTIC_SEARCH_PATH=/kibana/elasticsearch/logzioCustomerKibanaIndex
   - ELK_VERSION=6.2.1   KIBANA_TYPE=KibanaTypeVanilla
   - ELK_VERSION=6.3.2   KIBANA_TYPE=KibanaTypeVanilla
   - ELK_VERSION=6.4.1   KIBANA_TYPE=KibanaTypeVanilla

--- a/kibana/data_source_kibana_index_test.go
+++ b/kibana/data_source_kibana_index_test.go
@@ -68,7 +68,7 @@ func testAccDataSourceKibanaIndexLogz(dataSource string) resource.TestCheckFunc 
 
 		a := r.Primary.Attributes
 
-		expectedTitle := "[logz-"
+		expectedTitle := "[logzio"
 		if !strings.HasPrefix(a["title"], expectedTitle) {
 			return fmt.Errorf("expected kibana index title start with %s actual %s", expectedTitle, a["title"])
 		}

--- a/kibana/provider.go
+++ b/kibana/provider.go
@@ -4,83 +4,94 @@ import (
 	"fmt"
 	"log"
 	"os"
+	"sync"
 
 	"github.com/ewilde/go-kibana"
 	"github.com/hashicorp/terraform/helper/schema"
 	"github.com/hashicorp/terraform/terraform"
 )
 
+type KibanaProvider struct {
+	*schema.Provider
+	client *kibana.KibanaClient
+	once sync.Once
+}
+
 func Provider() terraform.ResourceProvider {
-	return &schema.Provider{
-		Schema: map[string]*schema.Schema{
-			"elastic_search_path": {
-				Type:        schema.TypeString,
-				Optional:    true,
-				DefaultFunc: envDefaultFuncWithDefault(kibana.EnvElasticSearchPath, kibana.DefaultElasticSearchPath),
-				Description: "The elastic search path, defaults to: " + kibana.DefaultElasticSearchPath,
+	kp := &KibanaProvider{
+		Provider: &schema.Provider{
+			Schema: map[string]*schema.Schema{
+				"elastic_search_path": {
+					Type:        schema.TypeString,
+					Optional:    true,
+					DefaultFunc: envDefaultFuncWithDefault(kibana.EnvElasticSearchPath, kibana.DefaultElasticSearchPath),
+					Description: "The elastic search path, defaults to: " + kibana.DefaultElasticSearchPath,
+				},
+				"kibana_uri": {
+					Type:        schema.TypeString,
+					Optional:    true,
+					DefaultFunc: envDefaultFuncWithDefault(kibana.EnvKibanaUri, kibana.DefaultKibanaUri),
+					Description: "The address of the kibana admin url, defaults to: " + kibana.DefaultKibanaUri,
+				},
+				"kibana_type": {
+					Type:        schema.TypeString,
+					Optional:    true,
+					DefaultFunc: envDefaultFuncWithDefault(kibana.EnvKibanaType, kibana.KibanaTypeVanilla.String()),
+					Description: "The type of the kibana either vanilla or logz.io, defaults to: " + kibana.KibanaTypeVanilla.String(),
+				},
+				"kibana_version": {
+					Type:        schema.TypeString,
+					Optional:    true,
+					DefaultFunc: envDefaultFuncWithDefault(kibana.EnvKibanaVersion, kibana.DefaultKibanaVersion),
+					Description: "The version of kibana being terraformed either 6.0.0 or 5.5.3, defaults to: " + kibana.DefaultKibanaVersion,
+				},
+				"kibana_username": {
+					Type:        schema.TypeString,
+					Optional:    true,
+					DefaultFunc: envDefaultFuncWithDefault(kibana.EnvKibanaUserName, ""),
+					Description: "The username used to connect to kibana",
+				},
+				"kibana_password": {
+					Type:        schema.TypeString,
+					Optional:    true,
+					DefaultFunc: envDefaultFuncWithDefault(kibana.EnvKibanaPassword, ""),
+					Description: "The password used to connect to kibana",
+				},
+				"logzio_client_id": {
+					Type:        schema.TypeString,
+					Optional:    true,
+					DefaultFunc: envDefaultFuncWithDefault(kibana.EnvLogzClientId, ""),
+					Description: "The logz.io client id used when connecting to logz.io",
+				},
+				"logzio_account_id": {
+					Type:        schema.TypeString,
+					Optional:    true,
+					DefaultFunc: envDefaultFuncWithDefault("LOGZIO_ACCOUNT_ID", ""),
+					Description: "The logz.io account id used when connecting to logz.io",
+				},
+				"logzio_mfa_secret": {
+					Type:        schema.TypeString,
+					Optional:    true,
+					DefaultFunc: envDefaultFuncWithDefault(kibana.EnvLogzMfaSecret, ""),
+					Description: "The logz.io MFA secret if the account has it enabled.",
+				},
 			},
-			"kibana_uri": {
-				Type:        schema.TypeString,
-				Optional:    true,
-				DefaultFunc: envDefaultFuncWithDefault(kibana.EnvKibanaUri, kibana.DefaultKibanaUri),
-				Description: "The address of the kibana admin url, defaults to: " + kibana.DefaultKibanaUri,
+
+			DataSourcesMap: map[string]*schema.Resource{
+				"kibana_index": dataSourceKibanaIndex(),
 			},
-			"kibana_type": {
-				Type:        schema.TypeString,
-				Optional:    true,
-				DefaultFunc: envDefaultFuncWithDefault(kibana.EnvKibanaType, kibana.KibanaTypeVanilla.String()),
-				Description: "The type of the kibana either vanilla or logz.io, defaults to: " + kibana.KibanaTypeVanilla.String(),
-			},
-			"kibana_version": {
-				Type:        schema.TypeString,
-				Optional:    true,
-				DefaultFunc: envDefaultFuncWithDefault(kibana.EnvKibanaVersion, kibana.DefaultKibanaVersion),
-				Description: "The version of kibana being terraformed either 6.0.0 or 5.5.3, defaults to: " + kibana.DefaultKibanaVersion,
-			},
-			"kibana_username": {
-				Type:        schema.TypeString,
-				Optional:    true,
-				DefaultFunc: envDefaultFuncWithDefault(kibana.EnvKibanaUserName, ""),
-				Description: "The username used to connect to kibana",
-			},
-			"kibana_password": {
-				Type:        schema.TypeString,
-				Optional:    true,
-				DefaultFunc: envDefaultFuncWithDefault(kibana.EnvKibanaPassword, ""),
-				Description: "The password used to connect to kibana",
-			},
-			"logzio_client_id": {
-				Type:        schema.TypeString,
-				Optional:    true,
-				DefaultFunc: envDefaultFuncWithDefault(kibana.EnvLogzClientId, ""),
-				Description: "The logz.io client id used when connecting to logz.io",
-			},
-			"logzio_account_id": {
-				Type:        schema.TypeString,
-				Optional:    true,
-				DefaultFunc: envDefaultFuncWithDefault("LOGZIO_ACCOUNT_ID", ""),
-				Description: "The logz.io account id used when connecting to logz.io",
-			},
-			"logzio_mfa_secret": {
-				Type:        schema.TypeString,
-				Optional:    true,
-				DefaultFunc: envDefaultFuncWithDefault(kibana.EnvLogzMfaSecret, ""),
-				Description: "The logz.io MFA secret if the account has it enabled.",
+
+			ResourcesMap: map[string]*schema.Resource{
+				"kibana_search":        resourceKibanaSearch(),
+				"kibana_visualization": resourceKibanaVisualization(),
+				"kibana_dashboard":     resourceKibanaDashboard(),
 			},
 		},
-
-		DataSourcesMap: map[string]*schema.Resource{
-			"kibana_index": dataSourceKibanaIndex(),
-		},
-
-		ResourcesMap: map[string]*schema.Resource{
-			"kibana_search":        resourceKibanaSearch(),
-			"kibana_visualization": resourceKibanaVisualization(),
-			"kibana_dashboard":     resourceKibanaDashboard(),
-		},
-
-		ConfigureFunc: providerConfigure,
 	}
+
+	kp.ConfigureFunc = kp.providerConfigure
+
+	return kp
 }
 
 func envDefaultFuncWithDefault(key string, defaultValue string) schema.SchemaDefaultFunc {
@@ -111,26 +122,34 @@ func GetEnvVarOrDefaultBool(key string, defaultValue bool) bool {
 	}
 }
 
-func providerConfigure(d *schema.ResourceData) (interface{}, error) {
-	config := &kibana.Config{
-		ElasticSearchPath: d.Get("elastic_search_path").(string),
-		KibanaBaseUri:     d.Get("kibana_uri").(string),
-		KibanaType:        kibana.ParseKibanaType(d.Get("kibana_type").(string)),
-		KibanaVersion:     d.Get("kibana_version").(string),
-	}
+func (p *KibanaProvider) providerConfigure(d *schema.ResourceData) (interface{}, error) {
+	var err error
 
-	client := kibana.NewClient(config)
-	client.SetAuth(authForContainerVersion[config.KibanaType](config, d))
-	client.Config.Debug = GetEnvVarOrDefaultBool("KIBANA_DEBUG", false)
-
-	if accountId, ok := d.GetOk("logzio_account_id"); ok && len(accountId.(string)) > 0 {
-		err := client.ChangeAccount(accountId.(string))
-		if err != nil {
-			return nil, err
+	p.once.Do(func() {
+		config := &kibana.Config{
+			ElasticSearchPath: d.Get("elastic_search_path").(string),
+			KibanaBaseUri:     d.Get("kibana_uri").(string),
+			KibanaType:        kibana.ParseKibanaType(d.Get("kibana_type").(string)),
+			KibanaVersion:     d.Get("kibana_version").(string),
 		}
+
+		client := kibana.NewClient(config)
+		client.SetAuth(authForContainerVersion[config.KibanaType](config, d))
+		client.Config.Debug = GetEnvVarOrDefaultBool("KIBANA_DEBUG", false)
+
+		if accountId, ok := d.GetOk("logzio_account_id"); ok && len(accountId.(string)) > 0 {
+			err = client.ChangeAccount(accountId.(string))
+			return
+		}
+
+		p.client = client
+	})
+
+	if err != nil {
+		return nil, err
 	}
 
-	return client, nil
+	return p.client, nil
 }
 
 var authForContainerVersion = map[kibana.KibanaType]func(config *kibana.Config, d *schema.ResourceData) kibana.AuthenticationHandler{

--- a/kibana/provider.go
+++ b/kibana/provider.go
@@ -14,7 +14,7 @@ import (
 type KibanaProvider struct {
 	*schema.Provider
 	client *kibana.KibanaClient
-	once sync.Once
+	once   sync.Once
 }
 
 func Provider() terraform.ResourceProvider {

--- a/kibana/provider_test.go
+++ b/kibana/provider_test.go
@@ -13,11 +13,11 @@ import (
 )
 
 var testAccProviders map[string]terraform.ResourceProvider
-var testAccProvider *schema.Provider
+var testAccProvider *KibanaProvider
 var testConfig *kibana.Config
 
 func init() {
-	testAccProvider = Provider().(*schema.Provider)
+	testAccProvider = Provider().(*KibanaProvider)
 	testAccProviders = map[string]terraform.ResourceProvider{
 		"kibana": testAccProvider,
 	}

